crates/precompile/src/bls12_381/blst.rs

```diff
@@ -314,3 +335,298 @@ pub(super) fn pairing_check(pairs: &[(blst_p1_affine, blst_p2_affine)]) -> bool
+    a_y_1: &[u8; FP_LENGTH],
+) -> Result<blst_p2_affine, PrecompileError> {
+    _extract_g2_input(a_x_0, a_x_1, a_y_0, a_y_1, false)
+}
+/// Extracts a G2 point in Affine format from the x and y coordinates.
+///
+/// Note: Coordinates are expected to be in Big Endian format.
+/// This function will perform a G2 subgroup check if `subgroup_check` is set to `true`.
+fn _extract_g2_input(
+    a_x_0: &[u8; FP_LENGTH],
+    a_x_1: &[u8; FP_LENGTH],
+    a_y_0: &[u8; FP_LENGTH],
+    a_y_1: &[u8; FP_LENGTH],
+    subgroup_check: bool,
+) -> Result<blst_p2_affine, PrecompileError> {
+    let out = decode_g2_on_curve(a_x_0, a_x_1, a_y_0, a_y_1)?;
+
+    if subgroup_check {
+        // NB: Subgroup checks
+        //
+        // Scalar multiplications, MSMs and pairings MUST perform a subgroup check.
+        //
+        // Implementations SHOULD use the optimized subgroup check method:
+        //
+        // https://eips.ethereum.org/assets/eip-2537/fast_subgroup_checks
+        //
+        // On any input that fail the subgroup check, the precompile MUST return an error.
+        //
+        // As endomorphism acceleration requires input on the correct subgroup, implementers MAY
+        // use endomorphism acceleration.
+        if unsafe { !blst_p2_affine_in_g2(&out) } {
+            return Err(PrecompileError::Other("Element not in G2".to_string()));
+        }
+    }
+    Ok(out)
+}
+
+/// Checks whether or not the input represents a canonical field element
+/// returning the field element if successful.
+///
+/// Note: The field element is expected to be in big endian format.
+pub(super) fn read_fp(input: &[u8; FP_LENGTH]) -> Result<blst_fp, PrecompileError> {
+    if !is_valid_be(input) {
+        return Err(PrecompileError::Other("non-canonical fp value".to_string()));
+    }
+    let mut fp = blst_fp::default();
+    // SAFETY: `input` has fixed length, and `fp` is a blst value.
+    unsafe {
+        // This performs the check for canonical field elements
+        blst_fp_from_bendian(&mut fp, input.as_ptr());
+    }
+
+    Ok(fp)
+}
+
+/// Extracts a scalar from a 32 byte slice representation, decoding the input as a Big Endian
+/// unsigned integer. If the input is not exactly 32 bytes long, an error is returned.
+///
+/// From [EIP-2537](https://eips.ethereum.org/EIPS/eip-2537):
+/// * A scalar for the multiplication operation is encoded as 32 bytes by performing BigEndian
+///   encoding of the corresponding (unsigned) integer.
+///
+/// We do not check that the scalar is a canonical Fr element, because the EIP specifies:
+/// * The corresponding integer is not required to be less than or equal than main subgroup order
+///   `q`.
+pub(super) fn read_scalar(input: &[u8]) -> Result<blst_scalar, PrecompileError> {
+    if input.len() != SCALAR_LENGTH {
+        return Err(PrecompileError::Other(format!(
+            "Input should be {SCALAR_LENGTH} bytes, was {}",
+            input.len()
+        )));
+    }
```

Comment:
We could move this check higher up and either change the input to `&[u8;SCALAR_LENGTH]` or add an assert in here